sporkfish/endgame_tablebase.py

```diff
@@ -24,6 +32,50 @@ def __init__(self, endgame_tablebase_path: Optional[str] = None):
         self.endgame_tablebase_path = endgame_tablebase_path
 
 
+class LilaTablebase:
+    base_url = "http://tablebase.lichess.ovh/standard?fen="
+
+    def query_dtz_bestmove(board_fen: str) -> tuple | None:
+        full_url = LilaTablebase.base_url + board_fen
+        try:
+            response = requests.get(full_url).json()
+            best_move = (
+                chess.Move.from_uci(response["moves"][0]["uci"])
+                if len(response["moves"]) > 0
+                else None
+            )
+            # the json from lila db gives us our best moves and probes opponent's dtz
+            # refer to https://github.com/lichess-org/lila-tablebase
+            # our dtz is the inversion of that
+            dtz = (
+                -opponent_dtz if (opponent_dtz := response["moves"][0]["dtz"]) else None
+            )
+            logging.info(
+                "Lila query succeeded. DTZ value retrieved: {dtz}. Best move: {best_move}"
+            )
+            return (dtz, best_move)
+        except ConnectionError as _:
+            logging.warning("Lila query failed, skipping")
+            return None
+
+    def query_dtz(board_fen: str) -> int | None:
+        full_url = LilaTablebase.base_url + board_fen
+        try:
+            response = requests.get(full_url).json()
+            # if we cant find dtz in our local db based on our board position after our legal move
+            # we query lila db to see if it exists
+            # the board position will be after our legal move and therefore dtz obtained is opponent's
+            # our dtz is the inversion of that
+            dtz = -opponent_dtz if (opponent_dtz := response["dtz"]) else None
+            logging.info(
+                "Lila query succeeded. DTZ value retrieved: {dtz}. Best move: {best_move}"
+            )
+            return dtz
+        except ConnectionError as _:
+            logging.warning("Lila query failed, skipping")
+            return None
+
+
 class EndgameTablebase:
```

Comment:
If I understand correctly what you're intending here is:
* Fetch local DTZ
* If its not None, do the local stuff.
* If it is None, query Lila tablebase - sending it to categorize similar our local logic. 
* Compare against Lila again (not sure why we need this one).

This should work but might not be optimal because:
1. Lots of Lila calls are made - one for every legal move on the board.                        
2. It's not clear that the local/Lila DTZ's mean exactly same thing due to rounding errors. See the long discussion here https://github.com/KYLChiu/sporkfish/pull/111#discussion_r1483705180 

I think we should stick with a simpler design I was alluding to originally. Just get the first move from the Lila JSON response. We don't really need to care about the DTZ score - just get the best move. This circumvents the two issues above because we only make one Lila API call + we don't care if the DTZ's are not fully compatible against whatever we have, because we don't fetch them at all.

Moreover the logic becomes easier to extend. I'd suggest that we only keep this PR to implement `LilaTablebase` class, which looks generally good overall. This is what is described in https://github.com/KYLChiu/sporkfish/issues/105.  In the next PR, my original thought was to make `EndgameTablebase` an abstract base class. Then we create a new class called `LocalTablebase` inheriting from `EndgameTablebase`, moving all logic of querying local DTZs to find the best move there. Finally we have `CompositeTablebase` which just queries `LocalTablebase` and `LilaTablebase`  in some order we decide.

Reply:
to clarify:
1. lila calls are only made when local DTZ returns none so it is not one for every legal move on the board. how many local DTZ couldnt be fetched usually?
2. comparing it again with Lila because getting best move via the iterative route (looping all legal moves) might come up with a different best move than querying Lila itself so just doing a last comparison

agree with the rounding errors bit and your overall design with the 2nd PR.

i will revert this back to simple implementation sadly 😒

Reply:
1. By default, most of the time local tablebase queries will come up empty - we don't provide the tablebase files upfront so it's up to the user to download them. 6/7 piece queries will practically all be from Lila because these reach up to 100 GB/18 TB in size and only dedicated users would bother to download them in the first place.

2. Fair enough - it gives us more control over which categories we like. If you'd like to prove me wrong and check that the calls are fast and unlikely to fail, happy to go through with your solution (with a bit of a cleaning up - separating classes etc).